### PR TITLE
Probe for compatible helper binaries on Windows

### LINKED ISSUE #14: Credential helpers that don't match on exact name cannot be used
## Description 

I was testing the SDK containers work with Google Cloud's Artifact Registry and found what I think is an issue with locating and executing credential helpers on Windows specifically.

The setup docs for Artifact Registry are as follows:

![image](https://user-images.githubusercontent.com/573979/197342601-57ef9805-0271-42ed-a082-af52f0a01cfd.png)

The key point is they have you run this command, which adds the following config to your config.json:

```json
"credHelpers": {
  "us-south1-docker.pkg.dev": "gcloud"
  }
```

The `gcloud` credential helper is added to your PATH, but on Windows it's actually in the form of a `cmd` script:

<details>
<summary>C:\Users\chethusk\AppData\Local\Google\Cloud SDK\google-cloud-sdk\bin\docker-credential-gcloud.cmd</summary>

```cmd
@echo off
rem Copyright 2017 Google Inc. All Rights Reserved.                                                                                                                                                                                                                                                                     SETLOCAL                                                                                                                                                                                                                                                                                                                SET CLOUDSDK_COMPONENT_MANAGER_DISABLE_UPDATE_CHECK=1

call gcloud.cmd auth docker-helper "%*"

%COMSPEC% /C exit %ERRORLEVEL%
```
</details>

Because the name of the credential helper doesn't match exactly what's on the PATH, and since the process is launched with UseShellExecute set to false (required because the streams are redirected), this credential helper cannot be used successfully.

It's easy to prove this out by changing the helper name in the configuration to `gcloud.cmd`, which makes things work as expected.

## Proposal

I propose that this library, on Windows specifically, should do a manual PATH probe to find different variations of the specified helper name. Proposed priority ordering might be:

* exact name match
* NAME.exe
* NAME.cmd
* NAME.bat

are there other similar extensions that should be taken into account?

Answer to the above: PATHEXT contains the executable file extensions.

### PR DESCRIPTION
**Pull request by baronfel**
Fixes #14 by explicitly probing PATH for executables on all operating systems. Additionally, on Windows the executable extensions specified by PATHEXT are taken into account.

This enables helpers that are shipped a non-suffixed executables on Windows to be invoked successfully.

### CODE REVIEW

**Comment by baronfel** [comment]
The justification for this kind of probing behavior is entirely compat with the golang-based Docker CLI behavior. The Docker CLI uses a [wrapper](https://github.com/docker/cli/blob/110c4d92b883357c9fb3edc344c4fbec5f77896f/vendor/github.com/docker/docker-credential-helpers/client/command.go#L32-L42) around the go `exec` package that enforces absolute paths for commands, and   the way this is done to by using [`LookPath`](https://cs.github.com/golang/go/blob/5b606a9d2b7649532fe25794fa6b99bd24e7697c/src/os/exec/lp_windows.go?q=PATHEXT#L66-L142) to do the PATH/PATHEXT-based expansions and lookups.

File: src/Valleysoft.DockerCredsProvider.Test/CredsProviderTests.cs [hunk 1]
```diff
@@ -490,12 +433,155 @@ public async Task ConfigFileDoesNotExist()
...
+        Mock<IFileSystem> fileSystemMock = new();
+        fileSystemMock
+            .WithFile(dockerConfigPath, dockerConfigContent)
+            .WithFile(Path.Combine(pathRoot, $"{fullHelperName}{pathExts[0]}"))
+            .WithFile(Path.Combine(pathRoot, $"{fullHelperName}{pathExts[1]}"));
+
+        Mock<IEnvironment> envMock = new();
+        envMock.WithSystemProfileFolder();
+        envMock.WithPath(new List<string> { pathRoot });
+        envMock.WithExecutableExtensions(pathExts);
```
Comment (mthalman) [inline src/Valleysoft.DockerCredsProvider.Test/CredsProviderTests.cs:528]:
```suggestion
        envMock.WithSystemProfileFolder()
            .WithPath(new List<string> { pathRoot })
            .WithExecutableExtensions(pathExts);
```

File: src/Valleysoft.DockerCredsProvider.Test/MockExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            .Returns(1);
+        return mock;
+    }
+
+    internal static Mock<IEnvironment> WithSystemProfileFolder(this Mock<IEnvironment> mock) {
+        mock.Setup(o => o.GetFolderPath(It.IsAny<Environment.SpecialFolder>())).Returns<Environment.SpecialFolder>(arg => new EnvironmentWrapper().GetFolderPath(arg));
+        return mock;
+    }
+
+}
```
Comment (mthalman) [inline src/Valleysoft.DockerCredsProvider.Test/MockExtensions.cs:87]: Add newline at end of file

File: src/Valleysoft.DockerCredsProvider.Test/XUnitAttributes.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+using Xunit;
+
+public sealed class WindowsFact : FactAttribute
+{
+    public WindowsFact() {
+        if(!RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) {
+            Skip = $"Test only runs on {OSPlatform.Windows}.";
+        }
+    }
+}
```
Comment (mthalman) [inline src/Valleysoft.DockerCredsProvider.Test/XUnitAttributes.cs:13]: Add newline at end of file

File: src/Valleysoft.DockerCredsProvider/NativeStore.cs [hunk 4]
```diff
@@ -60,12 +65,61 @@ public async Task<DockerCredentials> GetCredentialsAsync(string registry)
...
+            return path
+                .Split(Path.PathSeparator)
+                .Select(pathDir => this.CheckForCandidateOnPath(commandNameCandidates, pathDir))
+                .FirstOrDefault(candidate => candidate is not null);
+        } else {
+            return null;
+        }
+    }
+
+    private static List<T> Singleton<T>(T item) => new List<T>(1) { item };
```
Comment (mthalman) [inline src/Valleysoft.DockerCredsProvider/NativeStore.cs:84]: Please define this in a separate `ListHelper` class.

**Review by mthalman — APPROVED** [review]

File: src/Valleysoft.DockerCredsProvider.Test/CredsProviderTests.cs (hunk 1 above)
Comment (baronfel) [reply src/Valleysoft.DockerCredsProvider.Test/CredsProviderTests.cs]: Applied this in a couple other places I saw that I'd missed as well.

**Review by mthalman — APPROVED** [review]

**Pull request merged**